include/libtorrent/kademlia/node.hpp

```diff
@@ -90,6 +90,39 @@ struct udp_socket_interface
 	~udp_socket_interface() {}
 };
 
+struct TORRENT_EXTRA_EXPORT protocol_descriptor
+{
+	protocol_descriptor(udp protocol);
+
+	udp protocol() const;
+	char const* family_name() const;
+	char const* nodes_key() const;
+
+	bool is_v4() const;
+	bool is_v6() const;
+
+	bool is_native(udp::endpoint const& ep) const;
+	bool is_native(address const& addr) const;
+
+	static protocol_descriptor v4();
+	static protocol_descriptor v6();
+
+	operator udp() const { return m_data.protocol; }
+
+private:
+
+	struct descriptor_data
```

Comment:
but if you make ``protocol_descriptor`` impossible to construct, other than the two instances that need to exist, and you always pass it around by const reference, there shouldn't be any difference..